[css-conditional-5] Support `@supports font-format()`

nt1m opened the pull request:
#### c80f8136b208a12341b3fe2ded610d340330c7dd
<pre>
[css-conditional-5] Support `@supports font-format()`
<a href="https://bugs.webkit.org/show_bug.cgi?id=254679">https://bugs.webkit.org/show_bug.cgi?id=254679</a>
rdar://107381176

Reviewed by Brent Fulgham.

<a href="https://drafts.csswg.org/css-conditional-5/#typedef-supports-font-format-fn">https://drafts.csswg.org/css-conditional-5/#typedef-supports-font-format-fn</a>

Re-use list from `@font-face format()` parsing for `@supports font-face()`.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-L5-expected.txt:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::identMatchesSupportedFontFormat):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::consumeSupportsFunction):
(WebCore::CSSSupportsParser::consumeSupportsFeatureOrGeneralEnclosed):
(WebCore::CSSSupportsParser::consumeSupportsSelectorFunction):
(WebCore::CSSSupportsParser::consumeSupportsFontFormatFunction):
(WebCore::CSSSupportsParser::consumeConditionInParenthesis):
* Source/WebCore/css/parser/CSSSupportsParser.h:

Canonical link: <a href="https://commits.webkit.org/262305@main">https://commits.webkit.org/262305@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21c1aa1b238fe48e1ee0ef25395df2bf54c77f15

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1217 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1256 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1927 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/1060 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1168 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1268 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1290 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1230 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1193 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1093 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1801 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1084 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1121 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/1131 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1100 "Passed tests") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1154 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/119 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-L5-expected.txt:
```diff
@@ -1,5 +1,5 @@
 
-FAIL font-format() function accepts a known format assert_equals: expected true but got false
+PASS font-format() function accepts a known format
```
:-D

brentfulgham commented on Source/WebCore/css/parser/CSSSupportsParser.cpp:
```diff
@@ -147,13 +147,32 @@ CSSSupportsParser::SupportsResult CSSSupportsParser::consumeSupportsSelectorFunc
         if (range.peek().type() == FunctionToken && range.peek().functionId() == CSSValueSelector)
             return consumeSupportsSelectorFunction(range);
 
+        if (range.peek().type() == FunctionToken && range.peek().functionId() == CSSValueFontFormat)
```
It doesn't seem like this check is strictly needed, since you check this and return `Invalid` if it doesn't meet these requirements. Seems like you're doing the same work twice.

It seems like this would be entirely equivalent:
```
    return consumeSupportsFontFormatFunction(range);
```

Alternatively, if you think you will need to add more range checks in the future, maybe we should change the internal check to an ASSERT, since we expect the check to happen before entering the method.

brentfulgham commented on Source/WebCore/css/parser/CSSSupportsParser.cpp:
```diff
@@ -147,13 +147,32 @@ CSSSupportsParser::SupportsResult CSSSupportsParser::consumeSupportsSelectorFunc
+CSSSupportsParser::SupportsResult CSSSupportsParser::consumeSupportsFontFormatFunction(CSSParserTokenRange& range)
+{
+    if (range.peek().type() != FunctionToken || range.peek().functionId() != CSSValueFontFormat)
+        return Invalid;
```
You make this same check before entering the method. If you worry you might forget to check before entering this code, maybe it should just be an ASSERT?

nt1m committed:
[css-conditional-5] Support `@supports font-format()`
https://bugs.webkit.org/show_bug.cgi?id=254679
rdar://107381176

Reviewed by Brent Fulgham.

https://drafts.csswg.org/css-conditional-5/#typedef-supports-font-format-fn

Re-use list from `@font-face format()` parsing for `@supports font-face()`.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-L5-expected.txt:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::identMatchesSupportedFontFormat):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
(WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::consumeSupportsFunction):
(WebCore::CSSSupportsParser::consumeSupportsFeatureOrGeneralEnclosed):
(WebCore::CSSSupportsParser::consumeSupportsSelectorFunction):
(WebCore::CSSSupportsParser::consumeSupportsFontFormatFunction):
(WebCore::CSSSupportsParser::consumeConditionInParenthesis):
* Source/WebCore/css/parser/CSSSupportsParser.h:

Canonical link: https://commits.webkit.org/262305@main
